src/OpenApi/src/Services/OpenApiDocumentService.cs

```diff
@@ -50,28 +53,36 @@ internal sealed class OpenApiDocumentService(
     internal bool TryGetCachedOperationTransformerContext(string descriptionId, [NotNullWhen(true)] out OpenApiOperationTransformerContext? context)
         => _operationTransformerContextCache.TryGetValue(descriptionId, out context);
 
-    public async Task<OpenApiDocument> GetOpenApiDocumentAsync(CancellationToken cancellationToken = default)
+    public async Task<OpenApiDocument> GetOpenApiDocumentAsync(IServiceProvider scopedServiceProvider, CancellationToken cancellationToken = default)
     {
         // For good hygiene, operation-level tags must also appear in the document-level
         // tags collection. This set captures all tags that have been seen so far.
         HashSet<OpenApiTag> capturedTags = new(OpenApiTagComparer.Instance);
+        InitializeTransformers(scopedServiceProvider);
         var document = new OpenApiDocument
         {
             Info = GetOpenApiInfo(),
-            Paths = await GetOpenApiPathsAsync(capturedTags, cancellationToken),
+            Paths = await GetOpenApiPathsAsync(capturedTags, scopedServiceProvider, cancellationToken),
             Servers = GetOpenApiServers(),
             Tags = [.. capturedTags]
         };
-        await ApplyTransformersAsync(document, cancellationToken);
+        try
+        {
+            await ApplyTransformersAsync(document, scopedServiceProvider, cancellationToken);
+        }
+        finally
+        {
+            await FinalizeTransformers();
+        }
         return document;
     }
 
-    private async Task ApplyTransformersAsync(OpenApiDocument document, CancellationToken cancellationToken)
+    private async Task ApplyTransformersAsync(OpenApiDocument document, IServiceProvider scopedServiceProvider, CancellationToken cancellationToken)
     {
         var documentTransformerContext = new OpenApiDocumentTransformerContext
         {
             DocumentName = documentName,
-            ApplicationServices = serviceProvider,
+            ApplicationServices = scopedServiceProvider,
             DescriptionGroups = apiDescriptionGroupCollectionProvider.ApiDescriptionGroups.Items,
         };
         // Use index-based for loop to avoid allocating an enumerator with a foreach.
```

Comment:
I think `List<T>.Enumerator` is a struct?

Reply:
Nit: If we're micro-optimizing this loop, do we want to call `Count` once, up-front, rather than on every iteration?